### ISSUE TITLE
Neural networks as Ordinary Differential Equations

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1795]
Neural networks as Ordinary Differential Equations<br>
<br>
https://ift.tt/2ExuYwl<br>
<br>
Kevin Gibson<br>
<br>
<div><div>
<div>
<p>Recently I found a paper being presented at NeurIPS this year, entitled <a href="https://arxiv.org/abs/1806.07366">Neural Ordinary Differential Equations</a>, written by Ricky Chen, Yulia Rubanova, Jesse Bettencourt, and David Duvenaud from the University of Toronto. The core idea is that certain types of neural networks are analogous to a discretized differential equation, so maybe using off-the-shelf differential equation solvers will help get better results. This led me down a bit of a rabbit hole of papers that I found very interesting, so I thought I would share a short summary/view-from-30,000 feet on this idea.</p>
<p>Typically, we think about neural networks as a series of discrete layers, each one taking in a previous state vector <span>\( \mathbf{h}_n \)</span> and producing a new state vector <span>\( \mathbf{h}_{n+1} = F(\mathbf{h}_{n}) \)</span>. Here, let's assume that each layer is the same width (e.g. <span>\( \mathbf{h}_{n} \)</span> and <span>\( \mathbf{h}_{n+1} \)</span> have the same dimension, for every <span>\( n \)</span>). Note that we don't particularly care about what <span>\(F\)</span> looks like, but typically it's something like <span>\(F(x) = \sigma(\sum_{i}\theta_i x_i)\)</span>, where <span>\(\sigma\)</span> is an activation function (e.g. <a href="https://en.wikipedia.org/wiki/Rectifier_(neural_networks)">relu</a> or a <a href="https://en.wikipedia.org/wiki/Logistic_function">sigmoid</a>), and <span>\(\theta\)</span> is a vector of parameters we're learning. This core formulation has some problems - notably, adding more layers, while theoretically increasing the ability of the network to learn, can actually <em>decrease</em> the accuracy of it, both in training and test results.</p>
<p>This problem was adressed by <a href="https://arxiv.org/pdf/1512.03385.pdf">Deep Residual Learning</a>, from Kaiming He, Xiangyu Zhang, Shaoqing Ren, and Jian Sun out of Microsoft Reasearch. The idea in a nutshell is to learn a function of the difference between layers: <span>\( \mathbf{h}_{n+1} = F(\mathbf{h}_n) + \mathbf{h}_n \)</span>. In the paper, they show this simple transformation in what you're learning allows the networks to keep improving as they add more layers. To me, this reminds me of <a href="https://en.wikipedia.org/wiki/Delta_encoding">delta encoding</a>, in which you represent a stream of data as a series of changes from the previous state. This can make certain types of data much more suitable to compression (see, e.g. <a href="https://gafferongames.com/post/snapshot_compression/">this article</a> from Glenn Fiedler on compressing physics data to send over a network). It makes some kind of sense that if delta encoding can make data easier to compress, it could also make it easier to represent for a neural network.</p>
<h3>Euler's method and residual networks</h3>
<p>But how do residual networks relate to differential equations? Suppose we have some constant that we'll call <span>\( \Delta t \in \mathbb{R}\)</span>. Then we can write the state update of our neural network as</p>
<p><span>\[ \begin{aligned}
\mathbf{h}_{t+1} &amp;= F(\mathbf{h}_t) + \mathbf{h}_t \\ \\
&amp;= \frac{\Delta t}{\Delta t} F(\mathbf{h}_t) + \mathbf{h}_t \\ \\
&amp;= \Delta t G(\mathbf{h}_t) + \mathbf{h}_t
\end{aligned}
\]</span></p>
<p>where now we're learning <span>\( G(\mathbf{h}_t) = F(\mathbf{h}_t)/\Delta t + \mathbf{h}_t \)</span>. If you have experience with differential equations, this formulation looks very familiar - it is a single step of <a href="https://en.wikipedia.org/wiki/Euler_method">Euler's method</a> for solving ordinary differential equations. It seems this was first noticed by Weinan E in <a href="https://link.springer.com/article/10.1007/s40304-017-0103-z">A proposal on Machine Learning via Dynamical Systems</a>, and expanded upon by Yiping Lu et al. in <a href="https://arxiv.org/pdf/1710.10121.pdf">Beyond Finite Layer Neural Networks</a>. However, Lu et al. continue to treat the network as a series of discrete steps, and use a discrete solver with fixed timesteps to come up with a novel neural network architecture. The reason for this is that we need to be able to train the networks, and it's not really clear how to "learn" a differential system. Chen, Rubanova, Bettencourt and Duvenaud solve this problem by using some clever math which enables them to compute the gradients they need for backpropagation.</p>
<h3>Evaluating ODEs</h3>
<p>Before we get to that, let's look at what we're trying to solve. If we consider a layer of our neural network to be doing a step of Euler's method, then we can model our system by the differential equation</p>
<p><span>\[ \frac{d \mathbf{h}(t)}{dt} = G(\mathbf{h}(t), t, \theta) \]</span></p>
<p>Here we've made explicit <span>\(G\)</span>'s dependency on <span>\(t\)</span>, as well as some parameters <span>\(\theta\)</span> which we will train on. In this formulation, the output of our "network" is the state <span>\(\mathbf{h}(t_1)\)</span> at some time <span>\(t_1\)</span>. Therefore, if we know how to describe the function <span>\(G\)</span>, we can use any number of off-the-shelf ODE solvers to evaluate the neural network.</p>
<p><span>\[ \mathbf{h}(t_1) = \text{ODESolve}(\mathbf{h}(t_0), G, t_0, t_1, \theta) \]</span></p>
<p>There is a ton of research on different methods that can be used as our ODESolve function, but for now we'll treat it as a black box. What matters is that if you substitute in Euler's method, you get exactly the residual state update from above, with</p>
<p><span>\[ \text{ODESolve}(\mathbf{h}(t_0), G, t_0, t_1, \theta) = \mathbf{h}(t_0) + (t_1 - t_0)G(\mathbf{h}(t_0), t_0, \theta)\]</span></p>
<p>However, we don't need to limit ourselves to Euler's method, and in fact will do much better if we use more modern approaches.</p>
<h3>Training the beast</h3>
<p>So how to train it? Suppose we have a loss function</p>
<p><span>\[ L(h(t_1)) = L(\text{ODESolve}(\mathbf{h}(t_0), G, t_0, t_1, \theta))\]</span></p>
<p>To optimize <span>\(L\)</span>, we require gradients with respect to its parameters <span>\(\mathbf{h}(t)\)</span> (the state of our system at time <span>\(t\)</span>), <span>\(t\)</span> (our "time" variable, which is sort of a continuous analog to depth), and <span>\(\theta\)</span>, our training parameters. <a href="https://cs.stanford.edu/~ambrad/adjoint_tutorial.pdf">The adjoint method</a> describes a way to come up with this. The adjoint method is a neat trick which uses a simple substitution of variables to make solving certain linear systems easier. Part of the reason this paper grabbed my eye is because I've seen the adjoint method before, in a completely unrelated area: fluid simulation! In <a href="http://grail.cs.washington.edu/projects/control/fluidAdjoint.pdf">this</a> paper from McNamara et al., they make controlling a fluid simulation easier by using the adjoint method to efficiently compute some gradients with respect to user controlled parameters. That certainly sounds similar to our problem.</p>
<p>So what is the adjoint method? Suppose we have 2 known matrices <span>\(A\)</span> and <span>\(C\)</span>, and an unknown vector <span>\(\mathbf{u}\)</span> and we would like to compute a product</p>
<p><span>\[\mathbf{u}^{\intercal}B \text{ such that } AB=C\]</span></p>
<p>We could first solve the linear system to find the unknown matrix <span>\(B\)</span>, then compute the product, but solving the linear system could be expensive.
Instead, let's solve a different problem. Let's find a vector <span>\(\mathbf{v}\)</span> and compute</p>
<p><span>\[\mathbf{v}^{\intercal}C \text{ such that } A^{\intercal}\mathbf{v}=\mathbf{u}\]</span></p>
<p>We can show that these are in fact the same problem:</p>
<p><span>\[\mathbf{v}^{\intercal}C = \mathbf{v}^{\intercal}AB = (A^{\intercal}\mathbf{v})^{\intercal}B = \mathbf{u}^{\intercal}B\]</span></p>
<p>Through this transformation, we've reduced the problem from solving for a matrix, and reduced it to solving for a vector. This can be a big computational win! So how do we use it to train networks? I'm not going to go into the complete details here as it's slightly involved, but Appendix B in the paper has the complete derivation.
In brief, we define the adjoint state as</p>
<p><span>\[ a(t) = -\partial L / \partial \mathbf{h}(t) \]</span></p>
<p>And we can describe its dynamics via</p>
<p><span>\[ \frac{da(t)}{dt} = - a(t)^{\intercal} \frac{\partial G(\mathbf{h}(t), t, \theta)}{\partial \mathbf{h}} \]</span></p>
<p>We can compute the derivative of <span>\(G\)</span> with respect to <span>\(\mathbf{h}\)</span> already - we compute this gradient during backpropagation of traditional neural networks. With this, we can then compute <span>\(a(t)\)</span> by using another call to an ODE solver. There is one other derivative, <span>\(dL/d\theta\)</span>, that can be computed similarly. The paper shows that we can wrap up all of these ODE solves into a single call to an ODE solver, which computes all the necessary gradients for training the system.</p>
<h3>What's the point?</h3>
<p>Why do we want to do this? According to the paper, we're able to train a model with much less memory, with fewer parameters, and we are able to backpropagate more efficiently. All of these seem like good things! Modern ODE solvers are also adaptive, and can do more work only when needed to get an accurate solution. In the paper they show an experiment where the number of function evaluations that the ODE solver does increases with the number of training epochs - effectively, the system can quickly reach a rough solution, then take more time to refine the training.</p>
<hr>
<p><figure><img alt="illustration of an ode network compared to a residual network" title="An example from the paper showing how using an ode solver can adaptively evaluate the function. Circles represent function evaluations." src="https://rkevingibson.github.io/img/ode_networks_1.png"><figcaption>An example from the paper showing how using an ode solver can adaptively evaluate the function. Circles represent function evaluations.</figcaption></figure></p>
<hr>
<p>I think the most interesting aspect is that treating our system like a continuous time model, allows us to predict continuous time systems. They show a way to take data which arrives at arbitrary times, rather than at fixed intervals, and they can predict the output at arbitrary future times. They test this on fairly simple synthetic data, predicting trajectories of spirals, but get really nice results. I definitely want to see more of this type of work in the future, on larger real-world problems, to see how it does. Being able to draw on the &gt; 100 years of research in solving differential equations could be very useful for a young field like deep learning, and hey, I just find the math neat.</p>
<p><a href="https://rkevingibson.github.io/blog">Back to posts</a></p>
</div>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via 

     Kevin Gibson's personal site 

 <br>
<br>
December 18, 2018 at 03:47PM
